### PR TITLE
Assignment3 final

### PR DESCRIPTION
**Pull request by ds-georgiev**

### CODE REVIEW

File: src/main/java/globals/ColorThemes.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
+package globals;
+
+import classes.ColorTheme;
+import java.awt.*;
+
+public class ColorThemes {
```
Comment (MrFDMz) [inline src/main/java/globals/ColorThemes.java:6]: Way to go the extra mile!

File: src/main/java/globals/GlobalsViews.java [hunk 2]
```diff
@@ -1,6 +1,210 @@
...
+import classes.ColorTheme;
+import classes.SnippetState;
+
+import javax.imageio.ImageIO;
+import javax.swing.*;
+import java.awt.*;
+import java.io.IOException;
+import java.io.InputStream;
+
 public class GlobalsViews {
```
Comment (MrFDMz) [inline src/main/java/globals/GlobalsViews.java:13]: Good Work

**Review by MrFDMz — COMMENTED** [review]
Good job! Clearly you know what you are doing, the program is very complete and full of features, I agree with the changes that I have seen. There are maybe still some very large classes, but I can see that they are meant to be like that to cover all functionality. Once again I very much like the effort you put into the interface, I can guess most groups stuck to command line. Very complex, even (for me) hard to follow sometimes, impressive work, well done!

File: src/main/java/classes/SnippetHelper.java [hunk 3]
```diff
@@ -46,46 +58,60 @@ public void editSnippet(Snippet snippetToBeEdited){
     * Both methods will return otdered list with snippets
```
Comment (danieljmchugh) [inline src/main/java/classes/SnippetHelper.java:58]: typo

**Review by danieljmchugh — COMMENTED** [review]
Again you guys have done an incredible job, which makes my job in finding any flaws much harder. I really appreciate the spacing with the style, making this much easier to go though. Though the code at face value does seem more complex, i believe its fitting with the added functionality of the system. Once again, greate job guys!

**Comment by igormazurek** [comment]
I wrote it in the previous review and I have to do it again – a really good job guys. I like using JSON files instead of a database also the GSON library seems like a good choice. The option to select the color theme is also really nice, the only thing that I would add is a dark mode.

**Review by Xidot — COMMENTED** [review]
I didn't think it is possible to make readable code even more readable!
Great job guys, i hope you're proud of your own work!

I do not have any additional comments to be honest.

File: docs/assignment3.md [hunk 4]
```diff
@@ -1,110 +1,776 @@
...
+- Updated Class Diagram with appropriate private/public methods and/or variables
+
+You can see that the changes are made and are reflected in our code and Class Diagram. 
+
+Since Assigment 2 we have refactored quite a lot of classes in our code: 
+- The `Color` class now has changed and it is used now.
+- We use `resource` folder to read all of the media files
+- Using external files to keep the data
+- Using parameters from the terminal in order to point to a custom created folder to keep the files created
+- Accessing the `home.directory` of teh current logged in user in order to keep the needed files there if no folder path was specified
```
Comment (pgeusau) [inline docs/assignment3.md:31]: Well done, you correctly addressed the important feedback from assignment 2

File: docs/assignment3.md [hunk 5]
```diff
@@ -1,110 +1,776 @@
...
-| **Constraints**  | Any additional constraints that the application of the design pattern is imposing, if any |
-| **Additional remarks**  | Optional, only if needed |
+| **Design pattern**  | Singleton |
+| **Problem**  | `SettingsParser` class cannot be instantiated twice. If indeed we have two parsers reading and writing to file can be tricky and can lead to "deadlocks" or lead to functionality malfunction. Settings are important for our application since this is the place where the user can set its color theme, add categories adn set default programing language. We dont want to trouble the user experience.  |
+| **Solution**  | Implementing a singleton version of `SettingsParser` class will restrict it to be initialized only once. That will solve our problem since we have only one list in the `SettingsParser` since we will read/write to the file only once we will escape all kind of deadlocks or other problems already mentioned. |
+| **Intended use**  | The class will be instantiated in `Globals` where will be the entry point of the parser |
+| **Constraints**  | There are no constrains |
+| **Additional remarks**  |  |
+
```
Comment (pgeusau) [inline docs/assignment3.md:56]: Good!

File: docs/assignment3.md [hunk 6]
```diff
@@ -1,110 +1,776 @@
...
+We are aware that this is only one design pattern appearing twice, but because of the fact that we were too far with the code during Assignment 2 we were facing the challenge to refactor basically the whole logic of teh project. We reviewed our code and found some appropriate places to integrate two more design patterns: `Decorator Pattern` and `Observer Pattern`. Here is the way we should of approach it since the beginning.
```
Comment (pgeusau) [inline docs/assignment3.md:57]: SnippetParser and SettingsParser from the Singleton design pattern are very similar and explain your reasoning behind that. Lacking a clear motivation.

File: docs/assignment3.md [hunk 7]
```diff
@@ -1,110 +1,776 @@
...
+Where we should of use the `Observer Pattern`:\
+If you see our code you will notice that in  every `AllSnippetsForm`, `DashboardForm`, `SettingsForm` we use an `update()` method which will update the `DefaultListModel` or we are setting a new `DefaultListModel` model to the `JList`.\
+Using the `Observer Pattern` can improve the code and will make sure that all of this list updates will happen dynamically. This means that we do not need to worry about forgetting to update teh UI so the user can see the newly created, edited, deleted snippets. Furthermore, we can also include the filtering since the `JList` is affected there as well. Which means that in the `update()` method we need to update the `DefaultListModel` the same way we are doing it now. After some research we did online, we saw that the `DefaultListModel` is supposed to work as an `Observer` which means that at the moment we change the `List` the model should pick up and update the views in the `JList`.
+
+Not implementing the mentioned scenarios is a huge miss, but the refactoring we had to do was too big of a change and we did not want to risk it. But this can be seen as an improvement room for our code.
```
Comment (pgeusau) [inline docs/assignment3.md:67]: As you already mentioned, you're only using one design pattern without implementing the other (critical) design patterns. However, you give at least a trivial application of 1 design pattern with good descriptions. Good that you mentioned where you should use the other design patterns.

File: docs/assignment3.md [hunk 8]
```diff
@@ -1,110 +1,776 @@
...
+`private void handleAllSnippetsButton() `\
+Called when the all snippets button is clicked. It will go to the all snippets screen and it will set the `Globals.currentState = ApplicationState.STATE_ADD_SNIPPET;` and the snippet state to `Globals.currentSnippetState = SnippetState.SNIPPET_NORMAL;` since depending on the snippet state some visuals will change.
+
+`private void handleAddSnippetsButton()`
+Called when the add snippets button is clicked. It will go to the add snippets screen and it will set the `Globals.currentState = ApplicationState.STATE_ADD_SNIPPET;` and the snippet state to `Globals.currentSnippetState = SnippetState.STATE_ADD_SNIPPET;` since depending on the snippet state some visuals will change.
+
+`private void update()`\
+Used to update the Recently added list when an item is deleted.
+
+#### SttingsForm.java
```
Comment (pgeusau) [inline docs/assignment3.md:444]: *SettingsForm.java

File: docs/assignment3.md [hunk 9]
```diff
@@ -1,110 +1,776 @@
...
+
+#### SttingsForm.java 
+This class will paint the settings screen. In this screen, the user can choose the color theme and the default programing language. What is more, he/she can add the categories here as well in a comma-separated list. When pressing the save button the settings will be applied automatically and also will be parsed back to the `settigns.json` file. This will play the role of “saving” your settings for next time or until they change again.
+
+`private void handleSaveButtonClick()`\
+Applies and saves the settings in the `setigns.json` file and updates the current. Furthermore, if any changes are made in teh Color Theme or the Categories, all this settings will be available at the moment the user changes the screen
+
+`private void setCurrentElements() `\
+Setting the values of the elements of teh screen.
+
```
Comment (pgeusau) [inline docs/assignment3.md:452]: Well done overall, however, not all classes of the class diagram are aligned with the Java classes in the implementation. For example, 'GlobalsFilePath', also missing a description of this class. Good overall multiplicities and naming conventions. You're putting well the entities together that are likely to change together. Your classes should be as independent as possible from each other, this is not fully the case. Every class has one single responsibility. Good overall descriptions of the classes, just work more precisely. There are a lot of small sentences or grammar mistakes which makes it confusing sometimes.  AllSnippetsForm and ColorTheme classes are too big (God Classes). I'm missing clear motivation for public attributes and operations. Use enumerations if you already know that possible values are a finite set. Overall, good usage of UML constructs and it shows fairly realistic models of the system. The model is complete.

File: docs/assignment3.md [hunk 10]
```diff
@@ -1,110 +1,776 @@
...
+The SnippetParser will use the `Gson` object that is part of the `Gson` library and serialize or deserialize the snippets. The `gsonObject` uses the dateFormat object that will set the date format to the date when the snippet is formatted. Since we are sure the `snippets.json` file will contain a list with snippets we have to make sure that we note that to the `Gson` object in order to be able to parse the data properly. This is done by extracting the class type from the list using the `snippetListType` object.
+The `SettingsParser` has the same logic as the SnippetParser with the difference that we do not use the type object since the settings can be only one. The data is handled with the same `Gson` library.
```
Comment (pgeusau) [inline docs/assignment3.md:472]: I'm missing all the attributes in the object diagram. The object diagram should represent a snapshot of your system at a given time, this is not clear in your object diagram and description. Overall, the descriptions are good, the usage of UML constructs is good to a certain extend (very basic). However, the object diagram is partially complete.

File: docs/assignment3.md [hunk 11]
```diff
@@ -1,110 +1,776 @@
...
+
+
+In this state, though, we will focus more on the editing method. To begin with, through snippetHelper the program moves to the updateSnippets method. The program uses this method to update the passed object in the allSnippetsList and send it back to the SnippetParser. In the SnippetParser the now updated list is used in the update function. There a conversion takes place that translates the snippet object to JSON format. After this is done, the new JSON-format outcome is saved in a file. Finally, in order to see the changes the data needs to be updated. So, the program updates the data model in the All Snippet Screen and the updated snippet becomes visible.
+
```
Comment (pgeusau) [inline docs/assignment3.md:537]: Very good and detailed state diagrams. The majority of your events in the state diagram (represent calls to operations) are correctly represented in your class diagram. Make sure that all your activity names are verbs. Your implemented Java classes are aligned to the state diagram. Good usage of UML constructs and good realistic models of your system. The model is complete, however, there are several inconsistencies.

File: docs/assignment3.md [hunk 12]
```diff
@@ -1,110 +1,776 @@
...
+`Figure representing the Sequence Diagram 3`\
+![Image of Sequence Diagram 3](./diagrams/assignment3/seq_d_assginment3_scenario3.png)
+
```
Comment (pgeusau) [inline docs/assignment3.md:674]: Good overall sequence diagrams and descriptions. Missing the min/max number of iterations in the "loop" fragment. You should've used the alt fragment instead of calling it 'switch'. Make sure to use the 'strict' fragment when using atomic sequences. Different names for the objects in your sequence diagram as your instances of classes present in the class diagram, they should be aligned. The sequences of messages seem to be aligned with your java implementation, good job. Overall, good descriptions and model represents most things needed. Just some mistakes with the usage of UML constructs.

File: docs/assignment3.md [hunk 13]
```diff
@@ -1,110 +1,776 @@
...
+3. Open Terminal/PowerShell
+4. Navigate to the folder where you have download the `project.jar` file (can be the same folder as the one in step 1)
+5. Run the following command:\
+`java -cp project.jar Main --path "<FULL FOLDER PATH>"`\
+or\
+`java -cp project.jar Main -p "<FULL FOLDER PATH>"`<br><br>
+` <FULL FOLDER PATH>` is the place where you need to paste the full path you copied in step 2.
+
+***Video***\
+![Plase see our live demo (if video do not load it can be downloaded from the repository)](./video/demo_assignment3.mp4)
```
Comment (pgeusau) [inline docs/assignment3.md:773]: Very good implementation, also well aligned with your UML models. Java classes follow the basic rules by Ousterhout and Robillard. Just some big classes are present. The code is very neat and well readable, although some parts are quite complex. You implemented most features described in assignment 1.

**Review by pgeusau — APPROVED** [review]
